src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -27,219 +42,226 @@ export class CallbackWorkflow implements Workflow {
   }
 
   /**
-   * Get the document to add for speculative authentication. Is empty when
-   * callbacks are in play.
+   * Get the document to add for speculative authentication. This also needs
+   * to add a db field from the credentials source.
    */
-  speculativeAuth(): Promise<Document> {
-    return Promise.resolve({});
+  async speculativeAuth(credentials: MongoCredentials): Promise<Document> {
+    const document = startCommandDocument(credentials);
+    document.db = credentials.source;
+    return { speculativeAuthenticate: document };
   }
 
   /**
-   * Execute the workflow.
-   *
-   * Steps:
-   * - If an entry is in the cache
-   *   - If it is not expired
-   *     - Skip step one and use the entry to execute step two.
-   *   - If it is expired
-   *     - If the refresh callback exists
-   *       - remove expired entry from cache
-   *       - call the refresh callback.
-   *       - put the new entry in the cache.
-   *       - execute step two.
-   *     - If the refresh callback does not exist.
-   *       - remove expired entry from cache
-   *       - call the request callback.
-   *       - put the new entry in the cache.
-   *       - execute step two.
-   * - If no entry is in the cache.
-   *   - execute step one.
-   *   - call the refresh callback.
-   *   - put the new entry in the cache.
-   *   - execute step two.
+   * Execute the OIDC callback workflow.
    */
   async execute(
     connection: Connection,
     credentials: MongoCredentials,
-    reauthenticate = false
+    reauthenticating: boolean,
+    response?: Document
   ): Promise<Document> {
-    const request = credentials.mechanismProperties.REQUEST_TOKEN_CALLBACK;
-    const refresh = credentials.mechanismProperties.REFRESH_TOKEN_CALLBACK;
-
+    const requestCallback = credentials.mechanismProperties.REQUEST_TOKEN_CALLBACK;
+    const refreshCallback = credentials.mechanismProperties.REFRESH_TOKEN_CALLBACK;
+    // At minimum a request callback must be provided by the user.
+    if (!requestCallback) {
+      throw new MongoInvalidArgumentError(
+        'Auth mechanism property REQUEST_TOKEN_CALLBACK is required.'
+      );
+    }
+    // Look for an existing entry in the cache.
     const entry = this.cache.getEntry(
       connection.address,
       credentials.username,
-      request || null,
-      refresh || null
+      requestCallback,
+      refreshCallback || null
     );
+    let result;
+    // Reauthentication must go through all the steps again regards of a cache entry
```

Comment:
Might be a typo?

```suggestion
    // Reauthentication must go through all the steps again regardless of a cache entry
```

Reply:
I actually moved this comment down and changed the wording as the reauth behaviour changed slightly.